Add support for `@font-face src tech()` and  `@supports font-tech()` (252153)

vitorroriz opened the pull request:
#### ccfcd8af6c04a080d6350d17f33959d0a1ced378
<pre>
Add support for `@font-face src tech()` and  `@supports font-tech()` (252153)
<a href="https://bugs.webkit.org/show_bug.cgi?id=252153">https://bugs.webkit.org/show_bug.cgi?id=252153</a>
rdar://105665900

Reviewed by Myles C. Maxfield.

Implementing @font-face src tech() [1] and @supports font-tech() [2]
according to:

[1] <a href="https://www.w3.org/TR/css-fonts-4/#font-face-src-parsing">https://www.w3.org/TR/css-fonts-4/#font-face-src-parsing</a>
[2] <a href="https://drafts.csswg.org/css-conditional-5/#typedef-supports-font-tech-fn">https://drafts.csswg.org/css-conditional-5/#typedef-supports-font-tech-fn</a>

* LayoutTests/TestExpectations:
- We should pass this test now.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-tech-expected.txt:
- We pass all the subtests now, except for 2, since we don&apos;t support these
two font techonologies and techonologies are validated during parsing.

 * LayoutTests/platform/gtk/TestExpectations:
 * LayoutTests/platform/wpe/TestExpectations:
 - Skipping test for these platforms until webkit.org/b/256310 gets resolved.
 * LayoutTests/TestExpectations:
 * LayoutTests/imported/w3c/web-platform-tests/css/css-conditional/js/CSS-supports-L5-expected.txt:
 * LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-tech-expected.txt:
 * Source/WebCore/css/CSSFontFaceSrcValue.cpp:
 (WebCore::CSSFontFaceSrcResourceValue::CSSFontFaceSrcResourceValue):
 (WebCore::CSSFontFaceSrcResourceValue::create):
 (WebCore::CSSFontFaceSrcResourceValue::customCSSText const):
 (WebCore::CSSFontFaceSrcResourceValue::equals const):
 * Source/WebCore/css/CSSFontFaceSrcValue.h:
 (WebCore::cssTextFromFontTech):
 * Source/WebCore/css/CSSPrimitiveValueMappings.h:
 (WebCore::fromCSSValueID):
 * Source/WebCore/css/CSSValueKeywords.in:
 * Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
 (WebCore::CSSPropertyParserHelpers::consumeFontTech):
 * Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
 * Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
 (WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):
 * Source/WebCore/css/parser/CSSSupportsParser.cpp:
 (WebCore::CSSSupportsParser::consumeSupportsFunction):
 (WebCore::CSSSupportsParser::consumeSupportsFontTechFunction):
 * Source/WebCore/css/parser/CSSSupportsParser.h:
 * Source/WebCore/platform/graphics/FontCustomPlatformData.h:
 * Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
 (WebCore::FontCustomPlatformData::supportsTech):
 * Source/WebCore/platform/graphics/mac/FontCustomPlatformDataMac.cpp:
 (WebCore::FontCustomPlatformData::supportsTech):
 * Source/WebCore/platform/graphics/win/FontCustomPlatformDataWin.cpp:
 (WebCore::FontCustomPlatformData::supportsTech):
 * Source/WebCore/svg/SVGFontFaceUriElement.cpp:
 (WebCore::SVGFontFaceUriElement::createSrcValue const):

Canonical link: <a href="https://commits.webkit.org/263725@main">https://commits.webkit.org/263725@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d419867a1083e71dcc2e37434396153d3d97d3b1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5551 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5702 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5890 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7101 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5555 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5551 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5927 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5676 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/7118 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5654 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5715 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4967 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7137 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3193 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4990 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/12117 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5059 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5070 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6836 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5506 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4502 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4959 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1325 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9064 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5321 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -233,24 +233,30 @@ static RefPtr<CSSFontFaceSrcResourceValue> consumeFontFaceSrcURI(CSSParserTokenR
         // FIXME: We allow any identifier here and convert to strings; specification calls for certain keywords and legacy compatibility strings.
         auto args = CSSPropertyParserHelpers::consumeFunction(range);
         auto& arg = args.consumeIncludingWhitespace();
-        if (!args.atEnd())
```
ops, I shouldn't remove this. I'll fix it.

nt1m commented on Source/WebCore/css/CSSFontFaceSrcValue.cpp:
```diff
@@ -72,17 +72,18 @@ bool CSSFontFaceSrcLocalValue::equals(const CSSFontFaceSrcLocalValue& other) con
 }
 
-CSSFontFaceSrcResourceValue::CSSFontFaceSrcResourceValue(ResolvedURL&& location, String&& format, LoadedFromOpaqueSource source)
+CSSFontFaceSrcResourceValue::CSSFontFaceSrcResourceValue(ResolvedURL&& location, String&& format, Vector<FontTech> technologies, LoadedFromOpaqueSource source)
```
I think this creates a copy of the vector, not sure this is desirable.

vitorroriz replied:
I don't think so, I'm creating a rvalue reference when invoking this constructor so there will be 2 moves but no copy.

litherum replied:
Tim is right, this causes a copy. The receiver and the sender both have to cooperate to move

litherum replied:
Why not unify the format and the technologies into the same data structure?

litherum commented on Source/WebCore/css/CSSFontFaceSrcValue.h:
```diff
@@ -59,9 +60,54 @@ class CSSFontFaceSrcLocalValue final : public CSSValue {
     WeakPtr<SVGFontFaceElement, WeakPtrImplWithEventTargetData> m_element;
 };
 
+enum class FontTech : uint8_t {
```
Is this the right file to define this enum? I think there is a better file but I can’t remember off the top of my head what its name is.

litherum commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -2410,6 +2411,38 @@ template<> constexpr FontOpticalSizing fromCSSValueID(CSSValueID valueID)
     return FontOpticalSizing::Enabled;
 }
 
+template<> constexpr FontTech fromCSSValueID(CSSValueID valueID)
```
Can we not use abbreviations? We should spell out “technology”

nt1m replied:
I agree, this isn't a web-exposed enum, so we can spell it out as it is.

litherum commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -8213,6 +8214,35 @@ bool identMatchesSupportedFontFormat(CSSValueID id)
     >(id);
 }
 
+bool identMatchesSupportedFontTech(CSSValueID id)
```
This is platform specific and needs to be in a platform specific file

litherum reviewed (approved):
Do we have tests for the CSSOM?

litherum commented on Source/WebCore/platform/graphics/mac/FontCustomPlatformDataMac.cpp:
```diff
@@ -101,4 +102,20 @@ bool FontCustomPlatformData::supportsFormat(const String& format)
         || equalLettersIgnoringASCIICase(format, "woff"_s);
 }
 
+bool FontCustomPlatformData::supportsTech(const FontTech& tech)
```
`FontCustomPlatformDataMac.cpp` doesn't seem right, since this isn't Mac-specific. Can we create a new file at `platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp` instead?

litherum replied:
(Ideally this stuff would be in PAL, but that's probably out of scope for this patch...)

litherum commented on Source/WebCore/platform/graphics/win/FontCustomPlatformDataWin.cpp:
```diff
@@ -107,4 +108,10 @@ bool FontCustomPlatformData::supportsFormat(const String& format)
 
+bool FontCustomPlatformData::supportsTech(const FontTech& tech)
+{
+    // FIXME: define supported techonlogies for this platform (webkit.org/b/256310).
```
(typo)

I think we should populate this fully before landing the patch. The maintainers of the other ports probably can just tell you what this function should say. `return true` is harmful, as it will cause unrecognized fonts to be marked as recognized on those ports.

litherum commented on Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
```diff
@@ -163,4 +164,10 @@ bool FontCustomPlatformData::supportsFormat(const String& format)
 
+bool FontCustomPlatformData::supportsTech(const FontTech& tech)
+{
+    // FIXME: define supported techonlogies for this platform (webkit.org/b/256310).
```
(Ditto)

cdumez commented on Source/WebCore/css/CSSFontFaceSrcValue.cpp:
```diff
@@ -117,14 +118,25 @@ bool CSSFontFaceSrcResourceValue::customTraverseSubresources(const Function<bool
+                builder.append(", ");
+            builder.append(cssTextFromFontTech(m_technologies[i]));
+        }
+        builder.append(")");
```
builder.append(')');

cdumez commented on Source/WebCore/css/CSSFontFaceSrcValue.h:
```diff
@@ -59,9 +60,56 @@ class CSSFontFaceSrcLocalValue final : public CSSValue {
+    Invalid
+};
+
+inline String cssTextFromFontTech(FontTech tech)
```
Should return an ASCIILiteral. We don't need an unnecessary String allocation here.

cdumez commented on Source/WebCore/css/CSSFontFaceSrcValue.h:
```diff
@@ -59,9 +60,56 @@ class CSSFontFaceSrcLocalValue final : public CSSValue {
+    case FontTech::Variations:
+        return "variations"_s;
+    default:
+        return emptyString();
```
""_s

nt1m commented on Source/WebCore/css/parser/CSSSupportsParser.cpp:
```diff
@@ -178,6 +180,14 @@ CSSSupportsParser::SupportsResult CSSSupportsParser::consumeSupportsFontFormatFu
+{
+    ASSERT(range.peek().type() == FunctionToken && range.peek().functionId() == CSSValueFontTech);
+    auto techs = CSSPropertyParserHelpers::consumeFontTech(range, true);
+    return techs.isEmpty() ? Unsupported : Supported;
```
```suggestion
    auto supportedTechnologies = CSSPropertyParserHelpers::consumeFontTech(range, true);
    return supportedTechnologies.isEmpty() ? Unsupported : Supported;
```

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
```diff
@@ -233,6 +233,7 @@ static RefPtr<CSSFontFaceSrcResourceValue> consumeFontFaceSrcURI(CSSParserTokenR
         return nullptr;
 
     String format;
+    Vector<FontTech> techs;
```
```suggestion
    Vector<FontTechnology> supportedTechnologies;
```

vitorroriz committed:
Add support for `@font-face src tech()` and  `@supports font-tech()` (252153)
https://bugs.webkit.org/show_bug.cgi?id=252153
rdar://105665900

Reviewed by Myles C. Maxfield.

Implementing @font-face src tech() [1] and @supports font-tech() [2]
according to:

[1] https://www.w3.org/TR/css-fonts-4/#font-face-src-parsing
[2] https://drafts.csswg.org/css-conditional-5/#typedef-supports-font-tech-fn

* LayoutTests/TestExpectations:
- We should pass this test now.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-tech-expected.txt:
- We pass all the subtests now, except for 2, since we don't support these
two font techonologies and techonologies are validated during parsing.

 * LayoutTests/platform/gtk/TestExpectations:
 * LayoutTests/platform/wpe/TestExpectations:
 - Skipping test for these platforms until webkit.org/b/256310 gets resolved.
 * LayoutTests/TestExpectations:
 * LayoutTests/imported/w3c/web-platform-tests/css/css-conditional/js/CSS-supports-L5-expected.txt:
 * LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/parsing/font-face-src-tech-expected.txt:
 * Source/WebCore/css/CSSFontFaceSrcValue.cpp:
 (WebCore::CSSFontFaceSrcResourceValue::CSSFontFaceSrcResourceValue):
 (WebCore::CSSFontFaceSrcResourceValue::create):
 (WebCore::CSSFontFaceSrcResourceValue::customCSSText const):
 (WebCore::CSSFontFaceSrcResourceValue::equals const):
 * Source/WebCore/css/CSSFontFaceSrcValue.h:
 (WebCore::cssTextFromFontTech):
 * Source/WebCore/css/CSSPrimitiveValueMappings.h:
 (WebCore::fromCSSValueID):
 * Source/WebCore/css/CSSValueKeywords.in:
 * Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
 (WebCore::CSSPropertyParserHelpers::consumeFontTech):
 * Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
 * Source/WebCore/css/parser/CSSPropertyParserWorkerSafe.cpp:
 (WebCore::CSSPropertyParserHelpersWorkerSafe::consumeFontFaceSrcURI):
 * Source/WebCore/css/parser/CSSSupportsParser.cpp:
 (WebCore::CSSSupportsParser::consumeSupportsFunction):
 (WebCore::CSSSupportsParser::consumeSupportsFontTechFunction):
 * Source/WebCore/css/parser/CSSSupportsParser.h:
 * Source/WebCore/platform/graphics/FontCustomPlatformData.h:
 * Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
 (WebCore::FontCustomPlatformData::supportsTech):
 * Source/WebCore/platform/graphics/mac/FontCustomPlatformDataMac.cpp:
 (WebCore::FontCustomPlatformData::supportsTech):
 * Source/WebCore/platform/graphics/win/FontCustomPlatformDataWin.cpp:
 (WebCore::FontCustomPlatformData::supportsTech):
 * Source/WebCore/svg/SVGFontFaceUriElement.cpp:
 (WebCore::SVGFontFaceUriElement::createSrcValue const):

Canonical link: https://commits.webkit.org/263725@main
